llvm/lib/Target/WebAssembly/WebAssemblyInstrSIMD.td

```diff
@@ -46,10 +46,12 @@ defm "" : ARGUMENT<V128, v2i64>;
 defm "" : ARGUMENT<V128, v4f32>;
 defm "" : ARGUMENT<V128, v2f64>;
 
-// Constrained immediate argument types
+// Constrained immediate argument types. Allow any value from the minimum signed
+// value to the maximum unsigned value for the lane size.
 foreach SIZE = [8, 16] in
 def ImmI#SIZE : ImmLeaf<i32,
-  "return -(1 << ("#SIZE#" - 1)) <= Imm && Imm < (1 << ("#SIZE#" - 1));"
+  // -2^(n-1) <= Imm <= 2^n-1, avoiding UB when n == 64.
+  "return -(1 << ("#SIZE#" - 1)) <= Imm && Imm <= int64_t(uint64_t(1 << ("#SIZE#" - 1) << 1) - 1);"
```

Comment:
- Why do we need `1 << ("#SIZE#" - 1) << 1`? Can't we just use `1 << "#SIZE#"`?
- Can you elaborate what `uint64_t` and `int64_t` casting do and what is the UB about?
- Why do we need to consider `n  == 64`, give that this is under `foreach SIZE = [8, 16]`?

Reply:
Oh nice, I completely missed that it says SIZE can only be 8 or 16 right there. Simplifications incoming!

None of this matters any more, but to answer your questions:

> * Why do we need `1 << ("#SIZE#" - 1) << 1`? Can't we just use `1 << "#SIZE#"`?

Shifting equal to or greater than the number of bits is UB, so `1 << SIZE` would be UB if size were 64. The workaround is to do `1 << 63 << 1`.

> * Can you elaborate what `uint64_t` and `int64_t` casting do and what is the UB about?

Now that I think about this more, this was never necessary. I was concerned about UB due to underflow on the subtraction, but subtracting 1 from 0 is perfectly fine.